### PR TITLE
[Client] Java client

### PR DESCRIPTION
**Pull request by mahatoankitkumar**
## Problem
Java CAC Client

### CODE REVIEW

File: clients/java/src/main/java/CAC/ExperimentationClient.java
Comment (ayushjain17) [inline clients/java/src/main/java/CAC/ExperimentationClient.java:1]: we keep the two clients separately in different projects

File: clients/java/exp-client/gradle/libs.versions.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# This file was generated by the Gradle 'init' task.
+# https://docs.gradle.org/current/userguide/platforms.html#sub::toml-dependencies-format
+
+[versions]
+guava = "33.1.0-jre"
```
Comment (Datron) [inline clients/java/exp-client/gradle/libs.versions.toml:5]: Why do we import guava? Do we need this?

File: clients/java/cac-client/src/main/java/CAC/CacClient.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    public String getStringAndFree(Pointer ptr) {
+        if (ptr == null) {
+            return null;
+        }
+        try {
+            return ptr.getString(0);
+        } finally {
+            freeString(ptr);
+        }
+    }
```
Comment (Datron) [inline clients/java/cac-client/src/main/java/CAC/CacClient.java:126]: Can we write this better?

File: clients/java/cac-client/src/main/java/CAC/Client.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.concurrent.CountDownLatch;
+
+import jnr.ffi.Pointer;
+
+public class Client {
+
+    private static void callCacClient() {
+        String dylib = "cac_client";
+        File currentDir = new File(System.getProperty("user.dir"));
+        String libraryPath = currentDir.getParentFile().getParentFile().getParentFile() + "/target/debug";
```
Comment (Datron) [inline clients/java/cac-client/src/main/java/CAC/Client.java:14]: No hardcoding the path please, use env SUPERPOSITION_LIB_PATH

File: clients/java/cac-client/src/main/java/CAC/Client.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        System.out.println("CAC Client");
+
+        System.out.println("---------------------");
+
+        CacClient wrapper = new CacClient(libraryPath, dylib);
+
+        int newClient;
+        try {
+            newClient = wrapper.cacNewClient(tenant, 1, "http://localhost:8080");
```
Comment (Datron) [inline clients/java/cac-client/src/main/java/CAC/Client.java:27]: The URL should not be hardcoded, is this an example file?

File: clients/java/cac-client/src/main/java/CAC/Client.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
+package CAC;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+
+import jnr.ffi.Pointer;
+
+public class Client {
```
Comment (Datron) [inline clients/java/cac-client/src/main/java/CAC/Client.java:9]: Can we name this class better? It seems like a Client implementation

File: clients/java/exp-client/src/main/java/CAC/Client.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
+package CAC;
```
Comment (Datron) [inline clients/java/exp-client/src/main/java/CAC/Client.java:1]: Package should be different

**Review by Datron — CHANGES_REQUESTED** [review]

File: clients/java/exp-client/gradle/libs.versions.toml (hunk 1 above)
Comment (mahatoankitkumar) [reply clients/java/exp-client/gradle/libs.versions.toml:5]: Not sure, let me check.

File: clients/java/cac-client/gradle/libs.versions.toml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# This file was generated by the Gradle 'init' task.
+# https://docs.gradle.org/current/userguide/platforms.html#sub::toml-dependencies-format
+
+[versions]
+guava = "33.1.0-jre"
+junit-jupiter = "5.10.2"
+
+[libraries]
+guava = { module = "com.google.guava:guava", version.ref = "guava" }
+junit-jupiter = { module = "org.junit.jupiter:junit-jupiter", version.ref = "junit-jupiter" }
```
[the remark this answers is not in the document]
Comment (mahatoankitkumar) [reply clients/java/cac-client/gradle/libs.versions.toml:10]: will check

File: clients/java/cac-client/src/main/java/CAC/Client.java (hunk 4 above)
Comment (mahatoankitkumar) [reply clients/java/cac-client/src/main/java/CAC/Client.java:27]: Yes this is an example file.

**Comment by Datron** [comment]
Can you add java and gradle versions in flakes.nix as well?

**Review by Datron — APPROVED** [review]

**Review by ShubhranshuSanjeev — APPROVED** [review]

**Pull request merged**